User interface

ImFriedHof opened the pull request:
This marks the merge of the UserInterfaceBranch to the main branch. 

In this branch the UI was created and it has the following structure: 

1. A frame that holds both the GraphPanel and the ControlPanel. (The ratio between these two is managed by a GridBagLayout with weighting )

2. The GraphPanel is taken from Mr. Noll's Code which was used as a codebase for this project. The ControlPanel itself is a TabbedPane which is used to select between the parametric and the polynomial functions. 
The two Panels that are created by this TabbedPane are named PolynomialFunctionControl and ParametricFunctionControl. 

3. PolynomialFunctionControl: This Panel (in combination with the FunctionControlPanel) is the most complicated because it stacks Panel into Panel because it is the only way to set a working Layout. For visual help I added a screenshot of a sketch that I made.
![Bild_2023-06-30_172558174](https://github.com/kantholz44mm/PolynomialPlotter/assets/136808165/99dd3004-00f0-4837-96d8-22320e239938)

The AddButton in the PolynomialHelpPanel adds one instance of the FunctionControlPanel. This is currently capped by 3, but could be increased to a max of about 7. 
The calculateButton gets the Text of the InputField and returns the ID/count of the instance it came from.
The deleteButton deletes its Instance of FunctionControlPanel.


4. The ParametricFunctionControl: This panel faces the same complexity issues as the Polynomial ome but in smaller. Its purpose is to be a layout panel to the ParametricHelpPanel and the ParametricFunctionControlPanel which are glued to the bottom and center respectively. 
The ParametricHelpPanel is a smaller version of the PoynomialHelpPanel (it misses the AddButton). 

The ParametricFunctionControlPanel implements the actual function input. It holds two FunctionFields for the X and Y function, two boxes for setting the range (which are proteted against nonNumeric inputs exept null and ".")  and a CalculateButton and a DeleteButton. 

The CalculateButton gets the input from the user in the textFields and the DeleteButton deletes all user Input.

An idea of mine is to rename some of those classes for more intuitive understanding of its contents as well as to further define the differenes to other classes and their names. 

FunctionControlPanel --> PolynomialFunctionInput
ParametricFunctionControlPanel --> ParametricFunctionInput

ImFriedHof committed:
Setup

ImFriedHof committed:
Setup

ImFriedHof committed:
This commit contains all the Code written on the local repo for the UserInterface. It was written in a separate repository due to issues regarding the setup of this repository on a Laptop.

The code implements a new frame which holds both the GraphPanel and the ControlPanel.
The ControlPanel implements further panels for the input of polynomial and parametric functions which are called ParametricFunctionControl and PolynomialFunctionControl.

The actual Input is made in the FunctionInput class which is implemented in the PolynomialFunctionControl class.

Following Problems will have to be solved:

Almost all Functions and Methods are implemented in the Constructors of the related UI component. This causes issues regarding the visibility.

Also the GraphPanel has to be adjusted and connected to the Buttons, etc. available in the UI.

ImFriedHof committed:
Cleanup done to the PolynomialFunctionControl class.

Almost everything was put into methods which are called in the constructor.

Furthermore an ActionListener for the calculate button was implemented which is used to create a new instance if FunctionInput (the Label that holds all the necessary tools to enter a function). The maximum of these instances is set to 3, but can be increased to 9.

ImFriedHof committed:
Added a parameter ID to the functionInput. This is to be able to check from which instance of FunctionInput an Input is coming (Color/Calculate/Delete/Function).

A HelpPanel was added which holds the Reset,Screenshot and Help Button.

Another Panel was added tob split the space as needed between the FunctionInputPosition (functionInput as before only placed in this Panel) and the HelpPanel.

Last but not least our new Logo was added (See Discord).

ImFriedHof committed:
Added a parameter ID to the functionInput. This is to be able to check from which instance of FunctionInput an Input is coming (Color/Calculate/Delete/Function).

A HelpPanel was added which holds the Reset,Screenshot and Help Button.

Another Panel was added tob split the space as needed between the FunctionInputPosition (functionInput as before only placed in this Panel) and the HelpPanel.

Last but not least our new Logo was added (See Discord).

ImFriedHof committed:
Added a parameter ID to the functionInput. This is to be able to check from which instance of FunctionInput an Input is coming (Color/Calculate/Delete/Function).

A HelpPanel was added which holds the Reset,Screenshot and Help Button.

Another Panel was added tob split the space as needed between the FunctionInputPosition (functionInput as before only placed in this Panel) and the HelpPanel.

Last but not least our new Logo was added (See Discord).

ImFriedHof committed:
Last 3 Commits are basically the same. With only minimal changes.

There were some misunderstandings between me and intellij.

This Commit updates the Logo for our Project because small changes were made to improve readability.

ImFriedHof committed:
It is now possible to add and delete the instances of the FunctionControlPanel(aka the "Inputpackages"). The number of instances is still capped at 3 and has a minimum of 1 but this could be changed.

ImFriedHof committed:
Implemented the ParametricFunctionControl class.

It now holds the Function Input for parametric Functions.
It can also get  the range parameters and prevents the user from inputting anything other than numbers into the range parameters.
The delete button deletes all input the user has done

ImFriedHof committed:
The ParametricFunctionControl implements the FunctionInput (ParametricFunctionControlPanel) in a new class for visual clearness resulting out of the need for a new Panel around the Input. This panel is for Layout purposes.

Also multiple Action Listeners were added. Most of which are still empty and have to be connected to the various parts of the function calculation and painting.

ImFriedHof committed:
New file which I could not find in the unversioned files.
Belongs to the last commit.

kantholz44mm commented on .idea/misc.xml:
The .idea folder does not belong on the Git repo. It's currently included in the .gitignore file, please check if you got these changes.

ImFriedHof replied:
Used the answer you gave me. Please check if it worked correctly. From what I can tell its better now.
This of course applies to the other files as well. I will not comment under all of them. This one should suffice.

FabianRolfMatthiasNoll replied:
The .iml is still there

kantholz44mm commented on .idea/modules.xml:
Neither does this

kantholz44mm commented on .idea/vcs.xml:
Neither does this

kantholz44mm commented on PolynomialPlotter.iml:
Neither does this. Please add this to the .gitignore file.

FabianRolfMatthiasNoll replied:
Still in the Repo. Not solved

kantholz44mm commented on src/ControlPanel.java:
```diff
@@ -0,0 +1,30 @@
+
+    public class ControlPanel extends JTabbedPane {
+
+        ControlPanel() {
```
Missing explicit visibility (public, private, protected, etc.)

ImFriedHof replied:
Done

kantholz44mm commented on src/ControlPanel.java:
```diff
@@ -0,0 +1,30 @@
+
+        ControlPanel() {
+
+            this.setBackground(Color.red);
```
Fine for testing, but this should be changed before merging

ImFriedHof replied:
Done

kantholz44mm commented on src/Frame.java:
```diff
@@ -0,0 +1,53 @@
+import java.awt.*;
+
+public class Frame extends JFrame {
+    Frame() {
```
Missing explicit visibility as well

ImFriedHof replied:
Done

kantholz44mm commented on src/Frame.java:
```diff
@@ -0,0 +1,53 @@
+import javax.swing.event.ChangeListener;
+import java.awt.*;
+
+public class Frame extends JFrame {
```
The name 'Frame' is a bit redundant as is. Maybe change to something like 'GraphPlotterFrame' etc.

ImFriedHof replied:
Done

kantholz44mm commented on src/FunctionControlPanel.java:
```diff
@@ -0,0 +1,199 @@
+
+public class FunctionControlPanel extends JPanel {
+
+    Color[] Colors = {Color.RED, Color.BLUE, Color.YELLOW, Color.GREEN, Color.GRAY};
```
Ambiguous variable name. What colors? Maybe change to 'GraphColors' etc. Also, inconsistent naming, starting with capital letters for some and regular ones for most other variables. Also missing explicit visibility specifier

ImFriedHof replied:
Done

kantholz44mm commented on src/FunctionControlPanel.java:
```diff
@@ -0,0 +1,199 @@
+    JComboBox<Color> colorPicker = new JComboBox<>(Colors);   //Combo Box to pick the Color of the Function
+    GridBagLayout InputLayout = new GridBagLayout();
+    GridBagConstraints gbc = new GridBagConstraints();
+    private int count;
```
count of what?

ImFriedHof replied:
Changed to CountOfInstances

kantholz44mm commented on src/FunctionControlPanel.java:
```diff
@@ -0,0 +1,199 @@
+        this.functionInputPosition = functionInputPosition;
+        this.count = count;
+        this.polynomialFunctionControl = polynomialFunctionControl;
+        this.setBackground(Color.blue);
```
Same as red background color for control panel, should be changed before merging

ImFriedHof replied:
Done

kantholz44mm commented on src/FunctionControlPanel.java:
```diff
@@ -0,0 +1,199 @@
+        }
+    }
+
+    private class ColorListRenderer extends DefaultListCellRenderer {
```
Maybe we can make use of JColorChooser:
https://docs.oracle.com/javase/tutorial/uiswing/components/colorchooser.html

(Optional, nice to have, but not required.)

FabianRolfMatthiasNoll replied:
As far as i know he tried it but it made things complicated. Would put this maybe on hold for a future "feature"?

ImFriedHof replied:
Done

kantholz44mm commented on src/ParametricFunctionControl.java:
```diff
@@ -0,0 +1,77 @@
+
+    private class ParametricHelpPanel extends JPanel {
+
+        GridLayout ParaHelpLabelLayout = new GridLayout(1, 2);
```
missing explicit visibility specifier

ImFriedHof replied:
Done

kantholz44mm commented on src/ParametricFunctionControlPanel.java:
```diff
@@ -0,0 +1,215 @@
+    JTextField functionFieldX = new JTextField();
+    JTextField functionFieldY = new JTextField();
+    JTextField rangeStart = new JTextField();
+    JTextField rangeEnd = new JTextField();
```
all of these are missing explicit visibility specifiers

ImFriedHof replied:
Done

kantholz44mm commented on src/ParametricFunctionControlPanel.java:
does the input field for ranges also accept negative numbers?

ImFriedHof replied:
It does now.

kantholz44mm commented on src/PolynomialFunctionControl.java:
```diff
@@ -0,0 +1,120 @@
+public class PolynomialFunctionControl extends JPanel {
+
+    public static int count = 0;
+    JPanel functionInputPosition = new JPanel(); //Panel that holds the FunctionInput (for Layout purposes)
```
missing visibility specifier

ImFriedHof replied:
Done

kantholz44mm commented on src/PolynomialFunctionControl.java:
```diff
@@ -0,0 +1,120 @@
+            revalidate();
+            repaint();
+            count++;
+            System.out.println("Add");
```
please remove debug outputs before merging

ImFriedHof replied:
Done

kantholz44mm commented on src/PolynomialFunctionControl.java:
```diff
@@ -0,0 +1,120 @@
+
+    private class PolynomialHelpPanel extends JPanel {
+
+        GridLayout HelpLabelLayout = new GridLayout(2, 2);
```
missing explicit visibility specifier

ImFriedHof replied:
Done

kantholz44mm commented on src/PolynomialFunctionControl.java:
```diff
@@ -0,0 +1,120 @@
+            }
+        }
+
+        PolynomialHelpPanel() {
```
missing explicit visibility specifier

ImFriedHof replied:
Done

kantholz44mm reviewed (changes requested):
Works in general, but there are some things that should be fixed before merging

FabianRolfMatthiasNoll commented on src/ParametricFunctionControlPanel.java:
```diff
@@ -0,0 +1,215 @@
+        add(Y);
+
+        //Creates the TextField for the actual input
+        //functionFieldY.setBorder(BorderFactory.createEmptyBorder());
```
Is this comment still needed? Uncommented Code is generally not a good practice. Also please review all your comments if they are really needed and useful

ImFriedHof replied:
Done

FabianRolfMatthiasNoll commented on src/ParametricFunctionControlPanel.java:
```diff
@@ -0,0 +1,215 @@
+                super.replace(fb, offset, length, range, attrs);
+        }
+
+        public boolean isNumeric(String range) {
```
Regarding this function. You do not need to assign the result of parseDouble to a variable if you are not going to use it.
Also in the first if statment the string could be empty (.isEmpty exists) maybe it would be useful to check for that also. Also move the return true in the try block

ImFriedHof replied:
I removed the variable and moved the return statement.
But I do not understand the need to check for an empty String. This function only checks user input and the user can not input "empty". The only thing the user can do is delete the complete String written in it, which is handled by the replace function.

FabianRolfMatthiasNoll replied:
I didnt look completely at the code. What happens when a user inserts a empty string? Because if he does it is not null it just has the length of 0 so "". Does it get filterd beforehand?

FabianRolfMatthiasNoll replied:
Done

FabianRolfMatthiasNoll commented on src/Frame.java:
```diff
@@ -0,0 +1,53 @@
+public class Frame extends JFrame {
+    Frame() {
+        this.setDefaultCloseOperation(EXIT_ON_CLOSE);
+        this.setSize(1400,900); //Sets the Size in case the Window is not maximized
```
Please review all the comments if they are useful. Many times the code is explaining itself more than enough. In that case pls remove the comments

ImFriedHof replied:
Done

FabianRolfMatthiasNoll commented on src/FunctionControlPanel.java:
```diff
@@ -0,0 +1,199 @@
+        }
+    }
+
+    public class DeriveActionListener implements ActionListener {
```
Why are the other ActionListener private and this one public?

ImFriedHof replied:
Changed

FabianRolfMatthiasNoll commented on src/ParametricFunctionControl.java:
```diff
@@ -0,0 +1,77 @@
+        this.add(new ParametricHelpPanel(), BorderLayout.PAGE_END);
+    }
+
+    private class ParametricHelpPanel extends JPanel {
```
Shouldnt the HelpTabel the exact same for both Panels? This would result in a double implementation and double change if changes come up. Make it a separate class and use it in both Tabs if it has to be tab specific

ImFriedHof replied:
The problem is that the AddButton is not needed in the parametricFunction tab as there is only one function at a time. But it is necessary in the polynomialFunction tab.
You are right, it is a double implementation of the 3 other buttons, but the uses of the Buttons also change. For example the DeleteButton in the polynomialFunction tab is used for deleting an entire instance of the FunctionControlPanel.
In the parametricFunction tab it is used for removing all the users inputs out of the TextFields but it does not delete the TextFields itself.
There sure is a better way to implement this but I dont know how to do it.

FabianRolfMatthiasNoll replied:
Ok then we will leave it for now. But maybe (we still have time) can split it into two panels. A general Panel (Screenshot, Help etc.) and a specific function control panel (polynomial: Cam Reset, Add? / parametric etc)

FabianRolfMatthiasNoll replied:
Done

FabianRolfMatthiasNoll commented on src/PolynomialFunctionControl.java:
```diff
@@ -0,0 +1,120 @@
+
+    }
+
+    private class PolynomialHelpPanel extends JPanel {
```
Same thing as in the ParametricFunctionControl ^^

ImFriedHof committed:
.gitignore is now working

ImFriedHof committed:
GraphPlotterFrame:
Changed Frame to GraphPlotterFrame and fixed visibility issues.

ControlPanel:
Deleted size setters because they are not used anyway.

PolynomialFunctionControl:
Fixed visibility issues, deleted debugging prints and changed/deleted background color.

FunctionControlPanel:
Changed variable names, fixed visibility issues, deleted debugging prints and fixed an issue with the ColorPicker, which would not update the selected color.

 ParametricFunctionControl:
 Fixed visibility issues.

 ParametricFunctionControlPanel:
 Fixed visibility issues and deleted unused code.

In all changed files comments were deleted.

ImFriedHof committed:
Changed name of variable count to countOfInstances.

Added the possibility for entering a negative range and repositioned the return.

Deleted unused import.
Corrected Typo.

ImFriedHof committed:
Added the .iml to the gitignore.

ImFriedHof committed:
As the last addition to the gitignore seems to have no affect here is another approach.

ImFriedHof committed:
As the last addition to the gitignore seems to have no affect here is another approach.

ImFriedHof committed:
Moved the ColorPicker related classes and methods to a new class called. ComboBoxColorPicker.

FabianNoll1996 committed:
Applied gitignore

FabianRolfMatthiasNoll committed:
Merge branch 'main' into userInterface
